File: tracing/CppDemangler.node/addon.js
var addon = require('bindings')('addon.node')
function demangle(name) {
    return addon.demangle(name);
}

class CppDemangler {
    async demangle(mangledName) {
        return demangle(mangledName);
    }
    exit() {}
}
module.exports = CppDemangler;
module.exports.demangle = demangle;

function testInteractively() {
    const input = '_ZZZN7android4Hwc24impl8Composer7executeEvENK4$_15clINS_8hardware8graphics8composer4V2_15ErrorEbjNS5_8hidl_vecINS5_11hidl_handleEEEEEDaRKT_RKT0_RKT1_RKT2_ENKUlSG_SJ_E_clIS9_NS5_12MQDescriptorIjLNS5_8MQFlavorE1EEEEESD_SG_SJ_'
    const output = demangle(input);
    console.log(input);
    console.log(output);

    const input2 = '__on_dlclose';
    const output2 = demangle(input2);
    console.log(input2, output2);
}
if (require.main === module) {
    console.log('called directly');
    testInteractively();
}

File: tracing/CppDemangler.node/addon.cc
#include <string>
#include <node.h>
#include <cxxabi.h>

using namespace v8;

static std::string demangle(const char *mangled_name, bool quiet = true) {
  int status = 0;
  const char *realname = abi::__cxa_demangle(mangled_name, 0, 0, &status);
  std::string res;
  switch (status) {
  case 0:
    // if (quiet) {
    //   puts(realname);
    // } else {
    //   printf("%s  %s\n", realname, mangled_name);
    // }
    res = realname;
    break;
  case -1:
    printf("FAIL: failed to allocate memory while demangling %s\n",
           mangled_name);
    break;
  case -2:
    // printf("FAIL: %s is not a valid name under the C++ ABI mangling rules\n",
    //        mangled_name);
    res = mangled_name;
    break;
  default:
    printf("FAIL: some other unexpected error: %d\n", status);
    break;
  }
  free((void *)realname);
  return res;
}

void demangle(const FunctionCallbackInfo<Value> &args) {
  Isolate *isolate = args.GetIsolate();

  // Check the number of arguments passed.
  if (args.Length() != 1) {
    isolate->ThrowException(Exception::TypeError(
        String::NewFromUtf8(isolate, "Wrong number of arguments")
            .ToLocalChecked()));
    return;
  }

  // Check the argument types
  if (!args[0]->IsString()) {
    isolate->ThrowException(Exception::TypeError(
        String::NewFromUtf8(isolate, "Wrong arguments").ToLocalChecked()));
    return;
  }

  v8::String::Utf8Value str1(isolate, args[0]);
  std::string res = demangle(*str1);

  args.GetReturnValue().Set(
      String::NewFromUtf8(isolate, res.c_str()).ToLocalChecked());
}
void Init(Local<Object> exports) { NODE_SET_METHOD(exports, "demangle", demangle); }

NODE_MODULE(addon, Init)
